lib/Sema/CSApply.cpp

```diff
@@ -4406,7 +4415,101 @@ namespace {
       // key path.
       assert(!baseTy || baseTy->hasUnresolvedType()
              || baseTy->getWithoutSpecifierType()->isEqual(leafTy));
-      return E;
+
+      if (!isFunctionType)
+        return E;
+
+      // If we've gotten here, the user has used key path literal syntax to form
+      // a closure. The type checker has given E a function type to indicate
+      // this; we're going to change E's type to KeyPath<baseTy, leafTy> and
+      // then wrap it in a larger closure expression with the appropriate type.
+
+      // baseTy has been overwritten by the loop above; restore it.
+      baseTy = exprType->getAs<FunctionType>()->getParams()[0].getPlainType();
+
+      // Compute KeyPath<baseTy, leafTy> and set E's type back to it.
+      auto kpDecl = cs.getASTContext().getKeyPathDecl();
+      auto keyPathTy =
+          BoundGenericType::get(kpDecl, nullptr, { baseTy, leafTy });
+      E->setType(keyPathTy);
+      cs.cacheType(E);
+
+      // To ensure side effects of the key path expression (mainly indices in
+      // subscripts) are only evaluated once, we construct an outer closure,
+      // which is immediately evaluated, and an inner closure, which it returns.
+      // The result looks like this:
+      //
+      //     return "{ $kp$ in { $0[keyPath: $kp$] } }( \(E) )"
+
+      auto &ctx = cs.getASTContext();
+      auto discriminator = AutoClosureExpr::InvalidDiscriminator;
+
+      // The inner closure.
+      //
+      //     let closure = "{ $0[keyPath: $kp$] }"
+      auto closureTy =
+          FunctionType::get({ FunctionType::Param(baseTy) }, leafTy);
+      auto closure = new (ctx)
+          AutoClosureExpr(E, leafTy, discriminator, cs.DC);
```

Comment:
Why not just a regular implicit closure? I believe that by conversion autoclosure is not supposed to have any arguments.

Reply:
The problem is the `discriminator`. `ClosureExpr`s get numbered during parsing, and `AutoClosureExpr`s get numbered during a walk after type checking. The parsing context is already gone by the time this code is running, so the only way to correctly set the `discriminator` is via an `AutoClosureExpr`.

Reply:
Interesting, and there is no way to change that behavior to allow implicit closures? It seems a bit like a hack to use autoclosure here, because autoclosure is not supposed to have any parameters and used only in `@autoclosure` marked positions, so we are kind of stretching it's design here...

Reply:
Well, the superclass `AbstractClosureExpr` is what handles parameters, not either of the subclasses. The main difference between `ClosureExpr` and `AutoClosureExpr` is that `ClosureExpr` is supposed to be for explicit closures parsed from the source code, and `AutoClosureExpr` is supposed to be for implicitly created closures around an expression.

So, it's true, this is the only place where `AutoClosureExpr` will currently have parameters. But the design of the classes is such that I think we'd be stretching the design much more to try to use `ClosureExpr` here.

Reply:
Maybe it would make sense to add a separate sub-class of `AbstractClosureExpr` which deals with general cases of implicit closures? I'm just trying to think through ideas here since `AutoClosureExpr` just seems like a specialized use-case to me. Maybe it doesn't really matter that much anyway since the only code which has to deal with that in in SILGen... I can't think of any special cases in CSDiag which deal with expressions like that.

Reply:
At the very least comment associated with `AutoClosureExpr` has to be updated to cover new use-case.